### ISSUE TITLE
Compilation control for larger projects

### ISSUE THREAD

**Issue by cassio-tav** [frescobaldi/frescobaldi#1301]
This issue is a feature request (which I think is easy to implement, and very valuable for those who use LilyPond seriously).

### Context

Suppose I'm writing a suite with some number of movements, for whatever ensemble, and I need to generate these outputs:

* `suite.complete.midi`
* `suite.mov-?.midi` (where '?' stands for the movement number)
* `suite.complete.concert-pitch.pdf` (this is for my own revision and analysis)
* `suite.complete.transposition.pdf` (for the conductor and whoever else he appoints)
* `suite.mov-?.transposition.pdf`(where '?' stands for the movement number)
* `suite.complete.instrument-?.pdf`(where '?' stands for the part number — for the players)

In this situation, each instrument's musical content will have to be compiled several times to generate all the `*.pdf`, and a few more to generate the `*.midi` (given that I'm using `articulate.ly`). The whole business could get very ugly if I didn't organize my include hierarchy and rationalize the project tree (when the number of files in a project is large, a looping include structure may not be so simple to debug). So, here's a plausible solution, one that I use often:

* `project-root`
  * `includes` (the project's general settings, variable initialization and functions definitions)
    * `include-1.ily`
    * `include-2.ily`
    * ...
  * `ensemble` (the ensemble compilable files for the whole suite and the movements)
    * `suite.complete.concert-pitch.ly`
    * `suite.complete.transposition.ly`
    * `suite.mov-1.transposition.ly`
    * `suite.mov-2.transposition.ly`
    * ...
  * `parts` (the compilable files for the several parts, implicitly with transposition if that applies to the specific instrument)
    * `suite.complete.instrument-1.ly`
    * `suite.complete.instrument-2.ly`
    * ...
  * `music-content` (not meant to be compiled on their own; the most important and time-consuming part of the work has to be the one that happens here: music composition)
    * `mov-1`
      * `suite.mov-1.instrument-1.ly`
      * `suite.mov-1.instrument-2.ly`
      * ...
    * `mov-2`
      * `suite.mov-2.instrument-1.ly`
      * `suite.mov-2.instrument-2.ly`
      * ...
    * ...
  * `midi`
    * `suite.complete.midi`
    * `suite.mov-1.midi`
    * `suite.mov-2.midi`
    * ...
  * `sheet-music`
    * `suite.complete.concert-pitch.pdf`
    * `suite.complete.transposition.pdf`
    * `suite.mov-1.transposition.pdf`
    * `suite.mov-2.transposition.pdf`
    * ...
    * `suite.complete.instrument-1.pdf`
    * `suite.complete.instrument-2.pdf`
    * ...

### Description of the problem

Actually, it is possible (kind of) to compile such a project in Frescobaldi as it is: I just have to define a session to isolate me from other projects; in the session settings dialog, define the base directory as `/path/to/project-root/`; and add to the *project-specific include paths* list (1) `/path/to/project-root/includes` and (2) all the `/path/to/project-root/music-content/mov-?`. Then I have to make sure that the desired compilable is the active open file in the editor, and hit **Control+m**.

The problem is that, given that Frescobaldi will not let me define which files are actually compilable, it will end up trying to compile any and everything. And so it will scatter `.pdf` and `.midi` files all over the place — it's quite messy. Even when you are not using *automatic engraving* (which should work well for any project, by the way), it's almost impossible, after hours of work, not once to forget to switch back to some main file before hitting 'compile'. And it's very annoying, when you're on a deadline, trying to debug a complex set of content-files, not only to have to clean up all the directories by hand, but also to have to keep multiple compilable files open alongside the files you're actually debugging. I should point out that with the project design strategy I’ve been describing, the compilable files are written in the initial phase of the development process, not to be touched again until the very last phase, when the focus shifts from composition to presentation and instructions texts, and to layout adjustments. Note that one of the main benefits of such a strategy is to isolate the creative work in a way that for most of the project's duration you are free from all worries except the music you’re making.

This wreckless compiling behavior of Frescobaldi got so messy for me that at one point I actually tried to switch to **VScode**. I loathe Microsoft from the bottom of my gut! And Frescobaldi is better anyway in many important respects. So, I’ve since returned to Frescobaldi — but for a long while now I have not been using it to compile. In the beginning of a project I prepare a Makefile and from then on I let `make` do the compiling for me. The price I pay for this choice is that (1) I lose the power to click on the error log and be directed to the corresponding code line, and (2) after every compilation I have to reload the project, to give Frescobaldi access to the updated PDF and MIDI files. Not pretty.

### The proposed feature

I would call this the *"wise compilation feature"*. It would *warrant that Frescobaldi simply **never tries to compile** a file that just **was not meant to be compiled***. But I'm not asking the software to figure out by itself which those are; I'm hoping for a much simpler and more straight forward mechanism — I will describe it as I have imagined it, but you people know better, please be forgiving...

#### Required new session settings:

Some extra settings would have to be added, to be entered by the user in the **session settings** dialog:

1. a list for all the compilable files, with two items per line:
   1. a string to store the path to the compilable file,
   1. a pointer to a sub-ordinate list of strings for all the dependencies of the corresponding compilable file (in the session settings dialog, it could maybe show as a button widget to open a sub-ordinate dialog);
1. a pointer to indicate which file in the compilable files list will be compiled by default (I think it could appear in the session settings dialog as a selection widget by the list items themselves);
1. a boolean to enable/disable the option `always compile all the compilable files`;
1. a boolean to enable/disable the option `let the last compiled be default`;
1. a string to set the path to the directory in which to store the `*.midi` output files;
1. a string to set the path to the directory in which to store the `*.pdf` output files.
1. a string to set the path to the directory in which to store the `*.svg` output files.

#### Notes regarding this feature’s description

* It should be assumed that all the settings above can be left blank by the user — I believe that a session should be valid as long as it has a name and a root directory.
* for precision and clarity, I'll try to use a somewhat "formal" logical grammar, but since I'm not a programmer, except in LilyPond, I'll do so in natural language.
* I'll need to refer to the data structures above. In order to emphasize that this is what I’m doing, I'll make these references as if those data structures existed in the form of shell variables (`$name`). The names will be self-explanatory.
* During initialization, a copy of ´$default_compilable_file´ must be set as a working variable that may be written over; if a default is not configured, this variable should be made `nul`, or something else testable.

#### The desired behavior on a  "compile" command being issued
(E.g., when the user hits **Control+m**)

1. **if** no session is active or the `$compilable_files_list` is empty, **then** compile the active open file in the editor, as it is currently done, and **return** the compilation exit status.
1. **If** `$compilable_files_list` has only one `$file`, **then**:
   1. **if** any of the entries in `$file`'s `$dependencies_list` is a missing file, **then** handle missing dependencies error and **return** the error;
   1. **else**, compile `$file` and return  the compilation exit status.
1. **else if** `$compile_all` is true, **then**
     1. **for all** `file` **in** `$compilable_list`:
          1. **if** any of the entries in `$file`'s `$dependencies_list` is a missing file, **then** handle missing dependencies error and **return** the error, thus breaking the loop (this is because in the context of a single project with multiple output files, there will always be a lot of shared dependencies, so that there’s no point in going on);
          1. **else** (the dependencies are OK or not defined):
               * compile `$file`;
               * **if** any of `$midi_output_directory`, `$pdf_output_directory` or `$svg_output_directory` are defined, **then** move all the `*.midi`, `*.pdf` and `*.svg` to where they belong, along with any intermediary files (sometimes, when there is a conversion failure, these are left behind even if Frescobaldi is configured to erase them).
               * **if** the compilation was successful **then** continue the loop;
               * **else** handle error and **return** the compilation exit status (same reason as above);
1. **else** (`$compile_all` is false):
   1. **if** `$working_default_compilable_file` is `nul`, **then** make it point to the first file in `$compilable_files_list`;
   1. prompt the user to select the `$file_to_be_compiled`, but offering `$working_default_compilable_file` preselected in the menu, so that the user can confirm it simply by hitting **Enter**;
   1.  **if** `$let_last_compiled_be_default` is true, **then** make the `$working_default_compilable_file` point to whatever file the current `$file_to_be_compiled` points to;
   1. **if** any of the entries in `$file`'s `$dependencies_list` is a missing file, **then** handle missing dependencies error and **return** the error;
   1. **else**, compile `$file_to_be_compiled`;
   1. **if** any of `$midi_output_directory`, `$pdf_output_directory` or `$svg_output_directory` are defined, **then** move all the `*.midi`, `*.pdf` and `*.svg` to where they belong, along with any intermediary files (as explained above).
   1. **return** the exit status of the compilation.

#### The end

If any of the above needs clarification, I'll be pleased to have you people get in touch. Thanks.

**Comment by fedelibre** [comment]
> So, I’ve since returned to Frescobaldi — but for a long while now I have not been using it to compile. In the beginning of a project I prepare a Makefile and from then on I let `make` do the compiling for me. The price I pay for this choice is that (1) I lose the power to click on the error log and be directed to the corresponding code line, and (2) after every compilation I have to reload the project, to give Frescobaldi access to the updated PDF and MIDI files. Not pretty.

There's a discussion about the support for external build tools (like Make, Meson, etc.) on issue #1069 . I think it's something which could be added as extension.

**Comment by fedelibre** [comment]
> The problem is that, given that Frescobaldi will not let me define which files are actually compilable, it will end up trying to compile any and everything. And so it will scatter `.pdf` and `.midi` files all over the place — it's quite messy. Even when you are not using _automatic engraving_ (which should work well for any project, by the way), it's almost impossible, after hours of work, not once to forget to switch back to some main file before hitting 'compile'. And it's very annoying, when you're on a deadline, trying to debug a complex set of content-files, not only to have to clean up all the directories by hand, but also to have to keep multiple compilable files open alongside the files you're actually debugging.

There are a couple of issues here.

**1. How to determine which files are actually compilable**
A simple idea could be adopting the common practice of using the .ily file extension for files to be included only and never compiled. Then Frescobaldi could have a preference to disable the compilation of the .ily files, as discussed in issue #798 

A smarter method would be parsing the file and check the presence of any block (\score, \markup,...), obviously not in a variable, which produces some output.

**2. Master file is always opened**
I guess you're using the `%-*- master:` variable to tell Frescobaldi which is the compilable file.
When I open a file containing a master variable, also the master file is opened automatically. I agree it's very annoying. Perhaps this should be added as a separate issue.

**Comment by cassio-tav** [comment]
> There's a discussion about the support for external build tools (like Make, Meson, etc.) on issue #1069 . I think it's something which could be added as extension.

If that feature is implemented and it works seamlessly with the editor, it would do the trick for me...

**Comment by cassio-tav** [comment]
> A simple idea could be adopting the common practice of using the .ily file extension for files to be included only and never compiled. Then Frescobaldi could have a preference to disable the compilation of the .ily files, as discussed in issue #798

For me that would be sub-optimal, because I have other extension usage criteria (`.ily` in my environment is reserved to general definitions and functions with no music-content of its own)

> A smarter method would be parsing the file and check the presence of any block (\score, \markup,...), obviously not in a variable, which produces some output.

Yes, that would be the smartest *decide-for-the-user* solution (which has real value, I know). This would be a neat solution that would also do the trick for me. But I still think it is important to check dependencies *before calling LilyPond*. It spares the user from moments of unnecessary stress.

> I guess you're using the `%-*- master:` variable to tell Frescobaldi which is the compilable file.
> When I open a file containing a master variable, also the master file is opened automatically. I agree it's very annoying. Perhaps this should be added as a separate issue.

I think I couldn't follow you on that one...

**Comment by fedelibre** [comment]
> > I guess you're using the `%-*- master:` variable to tell Frescobaldi which is the compilable file.
> > When I open a file containing a master variable, also the master file is opened automatically. I agree it's very annoying. Perhaps this should be added as a separate issue.
> 
> I think I couldn't follow you on that one...

Which part exactly?

I'll be verbose...
As documented in [Document Variables](https://github.com/frescobaldi/frescobaldi/blob/master/frescobaldi_app/userguide/docvars.md), the master file variable allows to compile another LilyPond document instead of the current. This is very useful when you split a piece in two files: the actual compilable file containing the call to the \score (the *master document*); and the file containing only music content (let's call it the *content document*). Usually you'll work mostly on the latter,  that's why the master variable was conceived.

Now the bug, or annoying behaviour, is the following:

- if I open the content document, Frescobaldi will automatically open also the master document
- if I close the master document and compile the content document, again Frescobaldi will open the master document

It's annoying because I don't want to edit the master document.
If I'm working on several pieces, it's very annoying as browsing from tab to tab will take much more.

**Comment by cassio-tav** [comment]
> Now the bug, or annoying behaviour, is the following:
> 
> * if I open the content document, Frescobaldi will automatically open also the master document
> * if I close the master document and compile the content document, again Frescobaldi will open the master document
> 
> It's annoying because I don't want to edit the master document.
> If I'm working on several pieces, it's very annoying as browsing from tab to tab will take much more.

Right, Yes! That's part of my problem too. It was dumb of me not to get it right away -- your message was clear, sorry. It's because my thoughts were somewhere else: ↓ ↓ ↓

The `!master` variable doesn't suit me for other reasons as well, because there are several output files that I don't want to generate every time: when I'm working on the music content, I only want to generate the complete ensemble, concert-pitch `.pdf`. So, instead of having one single master that generates all outputs at once, I chose to have different masters to generate diferent output files, or diferent sets of files. Therefore, I end up having multiple masters to every content-file, which is not comfortable to manage.

Well, there would be, I guess, a workaround for that: I could have a single master-file and use scheme to conditionally generate outputs based on the content of a single variable that I can set according to what I intend to do at a given moment. I don't find that comfortable either. In the end it's better, and simpler, to use make, and to be able to get exactly the outputs I want by issuing commands such as `make all`, or `make woodwinds`, or `make concertpitch`, etc. Which is what I actually do.

To sum up, I don't use the `!master` variable because I don't currently use Frescobaldi to compile. And that's because, when I did use it, I wasn't able to satisfactorily get Frescobaldi to behave the way I needed it to.
